File: valid_perfect_square/extended.js
// Runtime 72 ms, Memory Usage 34.5 mb

function validSquare(num) {
    // n will be our counter. Starting at one will take care of edge cases.
    let n = 1
    // loop through while n squared is less than or equal to the input.
    while (n * n <= num){
        if (n * n === num){
            // if n squared equals the input, then the input is a valid square.
            return true
        } else if (n * n < num){
            // since we haven't found an answer yet, increase n.
            n++
        }
    }
    // if we make it through the loop, then we don't have a perfect square.
    return false
};

File: valid_perfect_square/extended.rb
# Runtime 52 ms, Memory Usage 9.3 mb

def is_perfect_square(num)
    # i will be our counter. starting with 1 helps with edge cases.
    i = 1
    # loop through while the square of our counter is less than our input.
    while (i ** 2 <= num)
        if i ** 2 == num
            # if i squared equals the input, we have a valid perfect square.
            return true
        else
            # if we don't have an answer yet, increase i.
            i += 1
        end
    end
    # if we made it through the loop, we know the input is not a perfect square.
    return false
end